### PR TITLE
#161967144 bookmark articles

### PR DESCRIPTION
**Pull request by denomash**
**What does this PR do?**
-  bookmarks articles

**Description of Task to be completed?**
- authenticated users can bookmark an article.
- authenticated users can un-bookmark an article.
- authenticated users can get all articles they have bookmarked.

**How should this be manually tested?**

- Clone the repo and cd into the project folder
- Create a virtual environment by running `python3 -m venv <name_of_environment>`
- Activate the virtual environment by running `source <name_of_environment>/bin/activate`
- Run `$ pip install -r requirements.txt`
- Then add the following environment variables to your computer:

 - `$ export DB_NAME='choose_a_dbname'`
 - `$ export DB_USER='db_username'`
 - `$ export DB_PASSWORD='db_role_password'`
 - `$ export DB_HOST='localhost or preffered host'`
 - `$ export DB_PORT='server_port_number'`
- `$ export SECRET_KEY='some_secret_key`
- `$ export SOCIAL_AUTH_FACEBOOK_KEY='<facebook_app_id>'`
- `$ export SOCIAL_AUTH_FACEBOOK_SECRET='<facebook_app_secret>'`
- `$ export SOCIAL_AUTH_TWITTER_KEY='<twitter_api_key>'`
- `$ export SOCIAL_AUTH_TWITTER_SECRET='<twitter_api_secret>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY='<google_client_id>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET='<google_client_secret>'`

- Change the current branch by running `git checkout  ft-bookmark-articles-161967144`
- Run `python3 manage.py makemigrations` then, `python3 manage.py migrate`
- Run `python3 manage.py runserver`

-  Bookmarking an article endpoint:
**POST** `article/<str:slug>/bookmark`

- Un-bookmarking an article endpoin:
**DEL** `article/<str:slug>/bookmark`

- Retriving all bookmarked articles endpoint:
**GET** `articles/all/bookmarks `

**What are the relevant pivotal tracker stories?**
- [#161967144](https://www.pivotaltracker.com/story/show/161967144/comments/197819922)

**Checklist (edited)**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

**Screenshots (if appropriate)**
**A user bookmarking or un-bookmarking an article that does not exist**
screenshot 2018-12-20 at 15 15 58

**A user bookmarking an article**
screenshot 2018-12-20 at 15 17 14

**A user bookmarking an article twice**
screenshot 2018-12-20 at 15 18 45

**A user un-bookmarking an article**
screenshot 2018-12-20 at 15 20 15

**A user un-bookmarking an article twice**
screenshot 2018-12-20 at 15 20 22

**A user retrieving bookmarked articles but has not bookmarked any **
screenshot 2018-12-20 at 15 23 41

**A user retrieving bookmarked articles**
screenshot 2018-12-20 at 15 24 46

**Bookmarked articles have a bookmarked fied for indicating if articles have been bookmarked**
screenshot 2018-12-20 at 18 58 28

### CODE REVIEW

File: authors/apps/bookmark/tests/test_bookmark.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            HTTP_AUTHORIZATION='Token ' + self.access_token
+        )
+
+        response2 = self.client.delete(
+            '/api/article/{slug}/bookmark'.format(slug=slug),
+            content_type='application/json',
+            HTTP_AUTHORIZATION='Token ' + self.access_token
+        )
+
+        self.assertIn(response.data['message'], 'Unbookmarking article {data} was successful'.format(data=slug))
```
Comment (bettblake08) [inline authors/apps/bookmark/tests/test_bookmark.py]: Define a function for these sets of assert statements. This is for good DRY practice.

**Review by bettblake08 — CHANGES_REQUESTED** [review]

File: authors/apps/bookmark/tests/test_bookmark.py (hunk 1 above)
Comment (denomash) [reply authors/apps/bookmark/tests/test_bookmark.py]: @bettblake08 I have tried to create a function but it does not simplify the code.

Comment (bettblake08) [reply authors/apps/bookmark/tests/test_bookmark.py]: It's ok then. We can leave it as it is.

**Review by bettblake08 — APPROVED** [review]

**Comment by Darius-Ndubi** [comment]
Kindly update the PR to show that the first request is a POST, not a GET

**Comment by Darius-Ndubi** [comment]
Kindly check on the response when unbook marking an article. I am getting this response 
screenshot 2018-12-20 at 19 10 38

**Review by davishooly — APPROVED** [review]

**Review by Darius-Ndubi — APPROVED** [review]

**Comment by Bryan-Cee** [comment]
Kindly check on the response that is given when one wants to view their bookmarked articles. It would be better if the whole article was returned and not just the slug.

**Review by Bryan-Cee — CHANGES_REQUESTED** [review]
Fix the response on retrieving bookmarked articles.

**Comment by denomash** [comment]
@Bryan-Cee We'll do that on the frontend. Slugs are more convenient since we shall append the url + slug

**Review by Bryan-Cee — APPROVED** [review]

**Review by jillo-abdullahi — APPROVED** [review]
Works fine 👍

**Pull request merged**